src/mqtt.client.ts

```diff
@@ -443,28 +461,28 @@ export class MqttClient<
         if (!this.autoReconnect)
             // this should never be true
             return false;
-        if (typeof this.autoReconnect === 'boolean') return !this.disconnected && this.ready && !this.connecting;
-
-        const base = this.autoReconnect.reconnectUnready || (!this.disconnected && this.ready && !this.connecting);
+        if (typeof this.autoReconnect === 'boolean') {
+            return this.autoReconnect;
+        }
 
-        return base && this.reconnectAttempt <= (this.autoReconnect.maxReconnectAttempts ?? 1);
+        return this.reconnectAttempt <= (this.autoReconnect.maxReconnectAttempts ?? 1);
     }
 
     protected async reconnect() {
         this.transport.reset();
         this.transformer = this.createTransformer();
         this.transformer.options.debug = this.transformer.options.debug ?? this.mqttDebug.extend('transformer');
-        await this.connect().catch(e => this.emitError(e));
+        return await this.connect();
     }
 
-    protected async setDisconnected(reason?: string): Promise<void> {
+    protected async setDisconnected(reason?: string | Error): Promise<undefined | Error | string> {
+        const willReconnect = this.shouldReconnect();
+        this.mqttDebug(`Disconnected. Will reconnect: ${willReconnect}. Reconnect attempt #${this.reconnectAttempt}`);
         this.reconnectAttempt++; // this should range from 1 to maxAttempts + 1 when shouldReconnect() is called
-        const willReconnect = this.autoReconnect && this.shouldReconnect();
-
         this.stopExecutingFlows(new AbortError('Client disconnected.'));
 
-        super.setDisconnected();
-        this.emitDisconnect(`reason: ${reason} willReconnect: ${willReconnect}`);
+        this._setDisconnected();
```

Comment:
Why is this prefixed? It should be fine to call the super-method. Or it would need to be renamed.

Reply:
Because of TypeScript error. It complained about the return-type. And there was 2 ways - either change super. setDisconnected signature or rename it. Changing signature is not right from the design perspective - parent method should not be changed just because child method needed it. The only right choice, in this case, is to rename.

Reply:
BTW MqttClient actually is not a child of MqttBase from the OOP point of view. `Base` is not an entity at its core. `MqttBase` is more of `MqttConnectionManager` or `MqttConnectionStateManager`, but with a leaked abstraction in the current implementation.  `MqttClient` should [contain it but not being inherited from it](https://en.wikipedia.org/wiki/Composition_over_inheritance).

The parent and child methods signatures incompatibility is a kinda litmus test showing this design doesn't reflect requirements.

Reply:
Oh sure